src/os/bluestore/BlueStore.cc

```diff
@@ -6364,6 +6388,74 @@ int BlueStore::_open_super_meta()
     dout(10) << __func__ << " bluefs_extents 0x" << std::hex << bluefs_extents
 	     << std::dec << dendl;
   }
+
+  // ondisk format
+  int32_t compat_ondisk_format = 0;
+  {
+    bufferlist bl;
+    int r = db->get(PREFIX_SUPER, "ondisk_format", &bl);
+    if (r < 0) {
+      // base case: kraken bluestore is v1 and readable by v1
+      dout(20) << __func__ << " missing ondisk_format; assuming kraken"
+	       << dendl;
+      ondisk_format = 1;
+      compat_ondisk_format = 1;
+    } else {
+      auto p = bl.begin();
+      try {
+	::decode(ondisk_format, p);
+      } catch (buffer::error& e) {
+      }
+      bl.clear();
+      {
+	r = db->get(PREFIX_SUPER, "compat_ondisk_format", &bl);
+	auto p = bl.begin();
+	try {
+	  ::decode(compat_ondisk_format, p);
+	} catch (buffer::error& e) {
+	}
+      }
+    }
+    dout(10) << __func__ << " ondisk_format " << ondisk_format
+	     << " compat_ondisk_format " << compat_ondisk_format
+	     << dendl;
+  }
+
+  if (latest_ondisk_format < compat_ondisk_format) {
+    derr << __func__ << " compat_ondisk_format is "
+	 << compat_ondisk_format << " but we only understand version "
+	 << latest_ondisk_format << dendl;
+    return -EPERM;
+  }
+  if (ondisk_format < latest_ondisk_format) {
+    int r = _upgrade_super();
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  _set_alloc_sizes();
+  return 0;
+}
+
+int BlueStore::_upgrade_super()
+{
+  dout(1) << __func__ << " from " << ondisk_format << ", latest "
+	  << latest_ondisk_format << dendl;
+  assert(ondisk_format > 0);
+  assert(ondisk_format < latest_ondisk_format);
+
+  if (ondisk_format == 1) {
+    // changes:
+    // - super: added ondisk_format
+    // - super: added min_readable_ondisk_format
+    // - super: added min_compat_ondisk_format
+    ondisk_format = 2;
+    _prepare_ondisk_format_super(t);
```

Comment:
this's fixed in the next commit but makes this one broken - hence might produce issues with bisect or cherry-pick...